src/na/device/Device.cpp

```diff
@@ -0,0 +1,499 @@
+    const size_t numParams, const double* params,
+    const QDMI_Operation_Property prop, const size_t size, void* value,
+    size_t* sizeRet) const -> int {
+  if ((sites != nullptr && numSites == 0) ||
+      (params != nullptr && numParams == 0) ||
+      (value != nullptr && size == 0) || prop >= QDMI_OPERATION_PROPERTY_MAX) {
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  ADD_STRING_PROPERTY(QDMI_OPERATION_PROPERTY_NAME, name.c_str(), prop, size,
+                      value, sizeRet)
+  ADD_SINGLE_VALUE_PROPERTY(QDMI_OPERATION_PROPERTY_PARAMETERSNUM, size_t,
+                            numParameters, prop, size, value, sizeRet)
+  if (type != Type::ShuttlingMove) {
+    ADD_SINGLE_VALUE_PROPERTY(QDMI_OPERATION_PROPERTY_DURATION, double,
+                              duration, prop, size, value, sizeRet)
+    ADD_SINGLE_VALUE_PROPERTY(QDMI_OPERATION_PROPERTY_FIDELITY, double,
+                              fidelity, prop, size, value, sizeRet)
+  }
+  if (!isShuttling(type)) {
+    ADD_SINGLE_VALUE_PROPERTY(QDMI_OPERATION_PROPERTY_QUBITSNUM, size_t,
+                              numQubits, prop, size, value, sizeRet)
+  }
+  return QDMI_ERROR_NOTSUPPORTED;
+}
+
+int MQT_NA_QDMI_device_initialize() {
+  std::ignore = qdmi::Device::get(); // Ensure the singleton is created
+  return QDMI_SUCCESS;
+}
+
+int MQT_NA_QDMI_device_finalize() { return QDMI_SUCCESS; }
+
+int MQT_NA_QDMI_device_session_alloc(MQT_NA_QDMI_Device_Session* session) {
+  return qdmi::Device::get().sessionAlloc(session);
+}
+
+int MQT_NA_QDMI_device_session_init(MQT_NA_QDMI_Device_Session session) {
+  if (session == nullptr) {
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  return session->init();
+}
+
+void MQT_NA_QDMI_device_session_free(MQT_NA_QDMI_Device_Session session) {
+  qdmi::Device::get().sessionFree(session);
+}
+
+int MQT_NA_QDMI_device_session_set_parameter(
+    MQT_NA_QDMI_Device_Session session, QDMI_Device_Session_Parameter param,
+    const size_t size, const void* value) {
+  if (session == nullptr) {
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  return session->setParameter(param, size, value);
+}
+
+int MQT_NA_QDMI_device_session_create_device_job(
+    MQT_NA_QDMI_Device_Session session, MQT_NA_QDMI_Device_Job* job) {
+  if (session == nullptr) {
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  return session->createDeviceJob(job);
+}
+
+void MQT_NA_QDMI_device_job_free(MQT_NA_QDMI_Device_Job job) { job->free(); }
+
+int MQT_NA_QDMI_device_job_set_parameter(MQT_NA_QDMI_Device_Job job,
+                                         const QDMI_Device_Job_Parameter param,
+                                         const size_t size, const void* value) {
+  if (job == nullptr) {
+    return QDMI_ERROR_INVALIDARGUMENT;
+  }
+  // The called function is only static because jobs are not supported. We keep
+  // the function call, however, as it is needed to free the job when the jobs
+  // are supported.
```

Comment:
This is just copy-pasted and needs to be adapted for the respective function. Also holds below.